<launch>
    <arg name="rival_name" default="rival"/>
    <arg name="side" default="0"/> <!-- 0 for yellow; 1 for blue -->

    <!-- EKF -->

    <node pkg="ekf" exec="ekf_node" name="ekf_node" output="screen">
        <param name="use_cam" value="0" />
        <param name="robot_parent_frame_id" value="map" />
        <param name="robot_frame_id" value="base_footprint" />
        <param name="/use_sim_time" value="false" />
        <param name="q_linear" value="1e-5" />
        <param name="q_angular" value="1e-6" />
        <param name="r_camra_linear" value="1e-2"  />
        <param name="r_camra_angular" value="0.15"  />
        <param name="r_threshold_xy" value="1e-3"    />
        <param name="r_threshold_theta" value="1e-2" />
    </node>


    <!-- lidar localization -->

    <node pkg="lidar_localization_pkg" exec="lidar_localization" name="lidar_localization" output="screen">
        <param name="side" value="$(var side)"/>
        <param name="debug_mode" value="False"/>
        <param name="visualize_candidate" value="False"/>
        <param name="likelihood_threshold" value="0.001"/>
        <param name="consistency_threshold" value="0.97"/>
    </node>

    <node pkg="tf2_ros" exec="static_transform_publisher" name="robot_to_laser" output="screen" args="--x 0 --y 0 --z 0 --roll 0 --pitch 0 --yaw 1.57 --frame-id /base_footprint --child-frame-id /laser" />
    <include file="$(find-pkg-share lidar_localization_pkg)/launch/rplidar_s3_launch.py" />
    <include file="$(find-pkg-share lidar_localization_pkg)/launch/obstacle_extractor_launch.xml" />


    <!-- local filter -->

    <include file="/home/user/localization-ws/src/localization-devel-ws/local/local_filter/launch/local_filter_whole.xml" />


    <!-- rival localization -->
    <group>
        <push-ros-namespace namespace="$(var rival_name)" />
            <node name="obstacle_detector_to_map" pkg="obstacle_detector" exec="obstacle_extractor_node">
                <param from="/home/user/localization-ws/src/localization-devel-ws/rival/rival_localization/config/rival_localization.yaml"/>
                <param name="frame_id" value="map"/>
                <remap from="raw_obstacles" to="/obstacles_to_map" />
                <remap from="scan" to="/scan" />
            </node>
     </group>
    <node pkg="rival_localization" exec="rival_localization" name="rival_localization" output="screen">
        <param from="/home/user/localization-ws/src/localization-devel-ws/rival/rival_localization/config/rival_localization.yaml"/>
        <remap from="raw_pose" to="/$(var rival_name)/raw_pose" />
        <remap from="final_pose" to="/$(var rival_name)/final_pose"/>
    </node>
</launch>